crates/language/src/language_settings.rs

```diff
@@ -35,13 +36,75 @@ pub fn init(cx: &mut AppContext) {
 }
 
 /// Returns the settings for the specified language from the provided file.
-pub fn language_settings<'a>(
+pub fn language_settings(
     language: Option<&Arc<Language>>,
     file: Option<&Arc<dyn File>>,
-    cx: &'a AppContext,
-) -> &'a LanguageSettings {
+    cx: &AppContext,
+) -> LanguageSettings {
     let language_name = language.map(|l| l.name());
-    all_language_settings(file, cx).language(language_name.as_deref())
+    let mut settings = all_language_settings(file, cx)
+        .language(language_name.as_deref())
+        .clone();
+    let path = file
+        .and_then(|f| f.as_local())
+        .map(|f| f.abs_path(cx))
+        .unwrap_or_else(|| std::path::PathBuf::new());
+    let mut cfg = ec4rs::properties_of(path).unwrap();
```

Comment:
We have to watch, otherwise, how can Zed update its settings accordingly and start formatting with different tab width or something else?

Reply:
I'm not sure I'm getting the use case. As of now I think we don't need to watch because we are checking for every `.editorconfig` file every time we need it! Have I missed a use case ?

Reply:
I was concerned that we store language settings in some corner of the code, and that's why we have
https://github.com/zed-industries/zed/blob/0fe22f2a48b03aac159fed44115c675f8566bec8/crates/editor/src/editor.rs#L11405 for cases when all settings are cached.

But I think I'm wrong and `LanguageSettings` retrieval follows this code path mostly, so never mind me.
But should we alter `AllLanguageSettings`-related retrieval too, as it is able to return `LanguageSettings` itself.

Reply:
I didn't see any problematic place related to `AllLanguageSettings` retrieval. It could be problematic though. But the earlier we do it, the worst the performance hit will be..

Also, someone hinted at having zed specific options in `.editorconfig` in the issue (e.g. `zed_soft_wrap` for instance). But I think we could postpone that.

Reply:
> It could be problematic though 

It is, of course, a problematic place since there are usages of the https://github.com/zed-industries/zed/blob/f1a2638d2912581542898218003491d4a7a4f17b/crates/language/src/language_settings.rs#L793-L800 method which may return incorrect settings without .editorconfig merged.

So all we did we have to move into that method instead.
But on the bright side, we have `self` there where we can place the information about paths and .editorconfig contents, improving our perf.

> the worst the performance hit will be

Which means we'll have to fix that that, after all.
It seems that we have a better chance now, since it's a method, not a function, and it has `&self` parameter in it — we can put more fields inside that and cache editorconfigs there.
(that's what I've tried to say in https://github.com/zed-industries/zed/pull/16349#discussion_r1719796019)

The only culprit we have is the fact that it's not `&mut self` in the API and while changing all usages seems possible, it's quite a large change.
So the way forward I imagine we can have have some `directories_with_editor_configs: Arc<parking_lot::RwLock<PathBuf, EditorConfigContent>>` field inside `AllLanguageSettings`, where `PathBuf` would be dirs' abs pathes.

All merging has to happen inside `pub fn language` of the same struct — this way we seem to proxy every possible settings access through the merging.

Then, per worktree, we will track FS changes as in

https://github.com/zed-industries/zed/blob/07d5e22cbe0bbec5f8ab0f478bf23cc76d7724f8/crates/project/src/project.rs#L8030-L8036

and there, we'll check change events for all kinds of projects (remote or not), for .editorconfig files.
We'll have to add some `AllLanguageSettings::get_global_mut(cx).insert/remove_editorconfig(...)` API and call it there.

We'll have to look at `ec4rs` closer and find a method that reads the config directly (could it be just https://docs.rs/ec4rs/latest/src/ec4rs/file.rs.html#97 ?) and read it during non-removed FS events.

Then, before merging, for every project path we get, we can just `pop()` its segments until we match all of the editorconfig files — those have to be merged, taking most recent config values.
There seems to be a logic for stopping the traversing but it looks very simple: https://spec.editorconfig.org/index.html#file-processing

Note that all this (merging, searching, etc.) is pure logic and definitely needs unit tests.

Reply:
As an alternative background watch approach, we can keep the recursive `ec4rs::properties_of` thing that will merge everything for us (hence no extra logic and unit tests are needed).

The background watch will give us a way to collect a list of directories with .editorconfig — so for every `AllLanguageSettings::language` call, we can do `ec4rs::properties_of` on a path unless before, we've already called `ec4rs::properties_of` on some other path that had the same sequence of `.editorconfig` files in the parent dirs.

If it has the same sequence of files and they were not changed since the previous resolution, it means that `ec4rs::properties_of` had already returned all correctly settings and we can reuse it.

I am not sure, but seems that we can organize it as `resolved_editorconfig_chains: RwLock<Vec<(Vec<PathBuf>, EditorConfigContent)>>` (all names and types are ad-hoc, think on it better) inside the `AllLanguageSettings`.
There also, we have to keep track of `dirs_with_editorconfig: HashSet<PathBuf>`.

Rust does not allow to use collections as keys in *Maps, so we'll use a Vec instead (as there's not that many directories with editorconfigs to worsen the perf here).

FS watches then will only remove things with `Vec::retain` in case of any editorconfig-related event — any added/changed/removed file may change `EditorConfigContent` resolution, if added between `HashSet<PathBuf>` paths or is any of those paths itself.
Also, the FS watches will update the `dirs_with_editorconfig` field.

And merging part will check that field for any new paths — for any path given, `dirs_with_editorconfig` will allow to build a chain of dirs with editorconfigs — if that matches anything inside `resolved_editorconfig_chains`, we can reuse it, otherwise back to recursive FS walk to cache a new value.